presto-raptorx/src/main/java/com/facebook/presto/raptorx/metadata/DatabaseMetadataWriter.java

```diff
@@ -0,0 +1,718 @@
+
+            // rollback master
+            dao.rollback(commit.getCommitId());
+
+            return commit;
+        });
+
+        // skip if no rollback needed
+        if (activeCommit == null) {
+            return;
+        }
+
+        long commitId = activeCommit.getCommitId();
+        RollbackInfo rollbackInfo = activeCommit.getRollbackInfo();
+
+        // rollback shards
+        for (Jdbi shardDbi : shardDbi) {
+            shardDbi.useTransaction(handle -> {
+                ShardWriterDao dao = handle.attach(ShardWriterDao.class);
+
+                // acquire shard lock
+                long abortedCommitId = getLockedAbortedCommitId(dao);
+
+                // skip if no rollback needed
+                if (abortedCommitId > commitId) {
+                    return;
+                }
+
+                dao.updateAbortedCommitId(commitId);
+
+                // rollback shard
+                dao.rollback(commitId);
+
+                for (long tableId : rollbackInfo.getWrittenTableIds()) {
+                    dao.rollbackCreatedIndexChunks(chunkIndexTable(tableId), commitId);
+                    dao.rollbackDeletedIndexChunks(chunkIndexTable(tableId), commitId);
+                }
+            });
+        }
+
+        // rollback indexes
+        for (long tableId : rollbackInfo.getCreatedTableIds()) {
+            shardDbi.get(getTableShard(tableId)).useHandle(handle ->
+                    dropIndexTable(handle, tableId));
+        }
+
+        for (TableColumn column : rollbackInfo.getAddedColumns()) {
+            shardDbi.get(getTableShard(column.getTableId())).useHandle(handle ->
+                    dropIndexTableColumn(databaseType, handle, column.getTableId(), column.getColumnId()));
+        }
+
+        // finish rollback
+        masterDbi.useTransaction(handle -> {
+            MasterWriterDao dao = handle.attach(MasterWriterDao.class);
+
+            // acquire global lock
+            getLockedCurrentCommitId(dao);
+
+            // verify we are still rolling back
+            ActiveCommit commit = dao.getActiveCommit();
+            if ((commit == null) || (commit.getCommitId() != commitId)) {
+                return;
+            }
+            verifyMetadata(commit.isRollingBack(), "Commit should be rolling back");
+
+            // finalize rollback
+            deleteActiveCommit(dao);
+        });
+    }
+
+    @Override
+    public long beginCommit()
+    {
+        return masterDbi.inTransaction(handle -> {
+            MasterWriterDao dao = handle.attach(MasterWriterDao.class);
+
+            // acquire global lock
+            long currentCommitId = getLockedCurrentCommitId(dao);
+
+            // verify no commit in progress
+            if (dao.getActiveCommit() != null) {
+                throw new PrestoException(RAPTOR_INTERNAL_ERROR, "Commit already in progress");
+            }
+
+            // allocate commit ID
+            long commitId = sequenceManager.nextValue("commit_id", 1);
```

Comment:
Related to the above question. Do we need a separate CommitID besides a TransactionID?